### DISCUSSION TITLE
[@types/node] v18.13.0 Breaking Change

### DISCUSSION THREAD

**Discussion by tpalacino** [DefinitelyTyped/DefinitelyTyped#64262] [Issues with a @types package]
Changes to the `package.json` in v18.13.0 break builds with semantic versions allowing any of the major version 18.

**The package.json changes causing the error:**
``` json
"typesVersions": {
    "<=4.8": {
        "*": [
            "ts4.8/*"
        ]
    }
},
```

**The build error:**
``` powershell
[15:31:42] [tsc] typescript version: 3.7.7
[15:31:58] Finished subtask 'tslint' after 16 s
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,42): error TS1005: ',' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,44): error TS1068: Unexpected token. A constructor, method, accessor, or property was expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,57): error TS1005: ';' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,65): error TS1005: ';' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,66): error TS1109: Expression expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,67): error TS1109: Expression expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1490,17): error TS1005: ',' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1490,26): error TS1005: ';' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1494,17): error TS1005: ',' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1494,26): error TS1005: ';' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1498,15): error TS1005: ';' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1498,41): error TS1005: '(' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1504,17): error TS1005: ',' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1504,32): error TS1005: ',' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1504,41): error TS1005: ';' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1504,47): error TS1109: Expression expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1508,17): error TS1005: ';' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1508,43): error TS1005: '(' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1512,26): error TS1005: ';' expected.
[15:32:07] Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1514,1): error TS1128: Declaration or statement expected.
[15:32:07] Error - 'tsc' sub task errored after 25 s
```

[a comment by an automated account is not shown]

**Reply by Kirnos** [reply]
Our builds started failing this morning because of this breaking change. Please fix or rollback ASAP.

**Comment by yoursunny** [comment]
As stated in [DefinitelyTyped README](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/395260243438dcf4799ba0117711cf2f11f48eec/README.md#support-window):

> Definitely Typed only tests packages on versions of TypeScript that are less than 2 years old.
> Currently versions 4.1 and above are tested.

The build error starts with:

> `[15:31:42] [tsc] typescript version: 3.7.7`

TypeScript 3.7 is no longer supported.
Please upgrade to TypeScript 4.9 and try again.

**Reply by tpalacino** [reply]
I understand what it says in the README. This is simple though. The only thing that changed between builds working and builds not working is the new **minor** version of `@types/node`. According to [Semantic Versioning](https://semver.org) the **major** version change indicates breaking changes. The update to this `package.json` caused breaking changes and did not increment the major version.

**Reply by Kirnos** [reply]
Introducing the breaking change in the Minor version update is against Semantic Versioning rules.

**Reply by Semigradsky** [reply]
It is not the breaking change since this version of TypeScript is not supported.

**Reply by tpalacino** [reply]
Well then, the `package.json` files for each **minor** version in v18.11.* need to be retro actively updated to prohibit the use of TypeScript version <4.2. Either way this package is the one that dropped the ball and it's causing problems. What a joke!

**Reply by tpalacino** [reply]
<sarcasm> what a great job this community software project does following the guidelines defined by the community. AWESOME!

**Reply by yoursunny** [reply]
TypeScript and DefinitelyTyped never claimed to follow SemVer though.
I've encountered breaking changes in `typescript` minor versions, multiple times.
You should pin to exact version, or at least a specific minor version.

**Reply by tpalacino** [reply]
NPMJS uses it. you publish your package there.

**Reply by tpalacino** [reply]
should I create a discussion and issue for getting the previous versions to break those builds too so everything work and doesn't consistently at least?

**Reply by tpalacino** [reply]
other people break the rules too isn't a valid argument. Also, I should be able to specify that I want version ^18.* and know that I won't have build problems because if there is a breaking change it'll be in v19 and NPM will not auto upgrade me to v19.*. It'll keep the version it downloads within the safety (supposedly) defined by the Semantic Versioning schema the package manager platform uses

**Reply by peterblazejewicz** [reply]
@tpalacino please try to understnad the DT approach is a balance between moving forward (providing best TS experience for packages that are not written as native TypeScript), driven by community effort of individuals, and taking care about backward compatible and frozen code base, e.g. stack with TS version that was 'current' some time ago. For this reason the situation like this happens from time to time, there was no wrong intent here, thx!

**Reply by tpalacino** [reply]
There was never a question of intent. I am/was not suggesting that someone adjusted the `package.json` to **intent**ionally break the build of dependent projects. I am suggesting that there was an oversight regarding the impact of certain changes. I'm no expert on the `@types/node` package but I can understand basic guidelines and the intent of those guidelines is to (a) allow the dependent project to continue to receive security and bug fixes automatically (since those would not require incrementing the major version) and (b) prevent this type of change that breaks things (AKA "breaking changes") for apparently no reason from happening.

When a person chooses to add a dependency on version 1.* of a package (understanding [the Semantic Versioning system that the package management platform employs](https://docs.npmjs.com/about-semantic-versioning)) they should expect everything to continue building indefinitely or until they choose to upgrade to another major version of that package.

This package did not abide by the version system employed by NPM and that broke the build of dependent projects.

It really isn't more complicated than that.

**Reply by Kirnos** [reply]
Let me give an example of the issue and its concequences of not using the Semantic Versioning rules.
One the dependencies deep in the dependency tree of our project is using @types/liftoff version 2.5.1 and this version is depending on @types/node: * - meaning that it always pulls the latest available version of @types/node. So this change broke other libraries and there is no way for us to lock the specific version.
We have multiple packages dependent on @types/node

Do you expect the developers to update all the packages in their project now?

https://www.npmjs.com/package/@types/liftoff/v/2.5.1?activeTab=explore
<img width="1072" alt="Screenshot 2023-02-07 at 15 58 53" src="https://user-images.githubusercontent.com/15732276/217365580-85915764-7a2c-45a1-9029-eaab0d2b78b0.png">

**Reply by yoursunny** [reply]
@tpalacino 
> I am suggesting that there was an oversight regarding the impact of certain changes.

DefinitelyTyped has GitHub Actions to ensure changes fall in the guidelines.
If you want to propose an oversight, such oversight must be implemented as GitHub Actions.
You can suggest such steps in GitHub Actions by opening a Pull Request.

---

@Kirnos 
> @types/liftoff version 2.5.1 and this version is depending on @types/node: *

If @types/node change breaks @types/liftoff directly, GitHub Actions would not have passed.
Thus, the issue is probably elsewhere.

> One the dependencies deep in the dependency tree of our project … and there is no way for us to lock the specific version.

Look into [npm overrides](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) and [yarn overrides](https://yarnpkg.com/configuration/manifest#resolutions) and [pnpm overrides](https://pnpm.io/package_json#pnpmoverrides).

> Do you expect the developers to update all the packages in their project now?

[Yes](https://dev.to/yoursunny/comment/16i64), you should upgrade all dependencies monthly.

**Reply by tpalacino** [reply]
I am not proposing that oversight be implemented. I was proposing that a lack of attention to detail resulted in an issue. Definition 2 from [this dictionary](https://www.dictionary.com/browse/oversight)

Regarding the invitation for me to submit a PR, I would respectfully decline to be associated with this project as a contributor of any level.

**Reply by jcesarmobile** [reply]
I have `"typescript": "~4.1.2"` and it's also breaking my project, so doesn't seem related to using typescript 3.7

**Reply by yoursunny** [reply]
@jcesarmobile
1. Check your `package-lock.json` `pnpm-lock.yaml` `yarn.lock`.
2. Ensure none of your (direct or indirect) dependencies is pulling in an unsupported TypeScript.
3. If you find any dependency pulling in unsupported TypeScript, open a bug in their repository and ask them to upgrade to latest TypeScript.

**Reply by jcesarmobile** [reply]
I don't have any other typescript versions in the `package-lock.json`

It works with typescript 4.2, but not with 4.1

**Comment by schapirama** [comment]
As I mentioned elsewhere, `cdktf` (and maybe JSII?) has `@types/node` as a dependency, and it's breaking with this latest change. 

```
$ node -v
    v18.14.0

    $ npm -v
    9.3.1

    $ tsc -v
    Version 4.9.5

    $ cdktf debug
    language: go
    cdktf-cli: 0.15.3
    node: v18.14.0
    cdktf: 0.15.3
    constructs: 10.1.243
    jsii: 1.74.0
    terraform: 1.3.7
    arch: arm64
    os: darwin 21.6.0
    go: go version go1.19.5 darwin/arm64    
    
    $ cdktf get
    ...
    Error: jsii compilation failed with non-zero exit code: 1
    [ERROR] jsii/compiler - Compilation errors prevented the JSII assembly from being created
    
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,42): error TS1005: ',' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,44): error TS1068: Unexpected token. A constructor, method, accessor, or property was expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,57): error TS1005: ';' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,65): error TS1005: ';' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,66): error TS1109: Expression expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1485,67): error TS1109: Expression expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1490,17): error TS1005: ',' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1490,26): error TS1005: ';' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1494,17): error TS1005: ',' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1494,26): error TS1005: ';' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1498,15): error TS1005: ';' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1498,41): error TS1005: '(' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1504,17): error TS1005: ',' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1504,32): error TS1005: ',' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1504,41): error TS1005: ';' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1504,47): error TS1109: Expression expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1508,17): error TS1005: ';' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1508,43): error TS1005: '(' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1512,26): error TS1005: ';' expected.
    Error - [tsc] node_modules/@types/node/ts4.8/util.d.ts(1514,1): error TS1128: Declaration or statement expected.
```

cc @ansgarm you may be interested

**Reply by yoursunny** [reply]
The problem is that [jsii 1.74.0 depends on TypeScript ~3.9](https://unpkg.com/browse/jsii@1.74.0/package.json), which is no longer supported by DefinitelyTyped.
Although you installed TypeScript 4.9 directly, jsii package is still trying to use the outdated version.
You should report a bug in jsii repository, asking them to upgrade to TypeScript 4.9.

**Reply by yoursunny** [reply]
https://github.com/aws/jsii/blob/7697571542df82b5d228faf85d0e637edb254519/package.json#L38

Latest revision of jsii is using TypeScript 4.7, which should be supported.
Thus, you can just install jsii from GitHub repository instead of the outdated version from NPM.

**Reply by yoursunny** [reply]
Hmm, jsii's use of TypeScript 3.9 had conflicts with DefinitelyTyped in the past too.

https://github.com/aws/jsii/issues/3609
https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/60310

**Reply by schapirama** [reply]
Thanks for the explanation, @yoursunny

**Comment by VeerJain89** [comment]
For now, we have added a temporary solution which is working on our development environments. You need to add below in tsconfig.json file:

`{
    "compilerOptions": {...
        "skipLibCheck": true
    ...}
}`

**Reply by tpalacino** [reply]
This is useful but may not work for everyone. We had this in our tsconfig already and still got broken by this minor version update. I'm not 100% sure but suspect it's because we also had `./node_modules/@types` listed explicitly in the `typeRoots`  and `node` listed explicitly in the `types` of our tsconfig

**Comment by wattachai** [comment]
`npm i -D @types/node@<target-version>` could be helpful as well. For example, if you need TS 3.9, go for `@types/node@17.0.41`.

**Reply by VeerJain89** [reply]
For full list of dependent @types/node versions as per TS, use below command.
`npm dist-tags @types/node`

**Reply by tpalacino** [reply]
This will also be helpful for people impacted by this breaking minor version update. This was one of my earliest troubleshooting steps. When I ran this command it did download the version I specified but did not convert the version to be explicit (the version still had the `^` character). So you may still need to manually remove the benefit of [Semantic Versioning](https://semver.org/) from your package.json and lock yourself in to a particular minor and patch

**Reply by tkaycodes** [reply]
Thank you! This fixed our issue

**Reply by c4b4d4** [reply]
This worked for my case. Thank you.

**Comment by Kirnos** [comment]
In our project the @types/node package is not used directly but utilized by multiple dependencies deep in the dependency tree so we needed to keep the right version in all of them.
Here is the solution that fixed this problem and allowed to lock the @types/node to the prior version 18.11.9:

1. Delete `node_modules` folder
2. Delete `package-lock.json` file
3. Update `package.json` manually - add the following section:
```
"resolutions": {
    "@types/node": "18.11.9"
},
```
4. Update`package.json` manually - add `preinstall` step to `scripts` section

```
"scripts": {
    "preinstall": "npm install --package-lock-only --ignore-scripts && npx npm-force-resolutions"
},
```

5. Run `npm install`
6. Build your project

**Reply by johnstonian** [reply]
this saved me hours of trial and error. thank you @Kirnos <3

**Comment by janoo1011** [comment]
![image](https://user-images.githubusercontent.com/12800687/221207296-17a03a8d-b2e3-4f51-9bee-04dd4fe3c615.png)
My angular project builds are failing due to @types. I tried all above steps suggested, still failing. Please help me resolve the issue.

**Reply by ShrihariMohan** [reply]
https://stackoverflow.com/questions/75386478/error-in-node-modules-types-node-ts4-8-util-d-ts148542-error-ts1005-ex 
As mentioned in this thread use `npm i -D @types/node@17.0.41`. This solved for us.

**Reply by janoo1011** [reply]
Hi, the issue happens while I'm running on AWS codebuild, however doesn't cause any issue while building in local. Can you let me know how I can there? Do you suggest I run this on buildspec.yaml ?

**Reply by ShrihariMohan** [reply]
Hey,  I suggest you to recreate the problem in local and solve it in local first. We're using bitbucket pipelines and aws code deploy. 

All we have to do was adding this in our package.json and build it.

**Reply by janoo1011** [reply]
Hi Can you please help me with what I need to be adding on package.json, because npm i -D @types/node@17.0.41 seems to be a command. Also where should I run the command.? Under scripts in package.json or in build spec.yaml? Can you share that piece of script?Have been stuck with this issue. Please help.

**Reply by ShrihariMohan** [reply]
In your `package.json` under `devDependencies` You have to add this 
`"@types/node": "^17.0.41",`

**Reply by tpalacino** [reply]
Can you show the build log output that indicates the tsc version being used?

**Reply by DaniyalAlam** [reply]
I had the same problem. i was using an older version of `typescript` dependency. Upgrading it from `4.0.5` to `4.6.3` solved the errors. I don't know why it started to fail overnight as it was working before.

**Comment by vikaskumar27071990** [comment]
**This worked for me**

i treld almost all ways the i just jpdate the version so it worked

npm install typescript@latest

TRY THE ABOVE COMMAND

**Reply by BrentBWu** [reply]
Thanks, this worked for me!
